lib/AST/DocComment.cpp

```diff
@@ -426,24 +426,24 @@ class CommentProviderFinder final {
     return std::nullopt;
   }
 
+  /// Check if there is an inherited protocol that has a default implementation
+  /// of `VD` with a doc comment.
   std::optional<ResultWithDecl> findDefaultProvidedDecl(const ValueDecl *VD) {
-    // Only applies to protocol extension member.
-    auto *protocol = VD->getDeclContext()->getExtendedProtocolDecl();
-    if (!protocol)
+    NominalTypeDecl *nominalType =
+        dyn_cast_or_null<NominalTypeDecl>(VD->getDeclContext()->getAsDecl());
+    if (!nominalType) {
+      nominalType = VD->getDeclContext()->getExtendedProtocolDecl();
+    }
+    if (!nominalType)
       return std::nullopt;
 
     SmallVector<ValueDecl *, 2> members;
-    protocol->lookupQualified(const_cast<ProtocolDecl *>(protocol),
-                              DeclNameRef(VD->getName()),
-                              VD->getLoc(), NLOptions::NL_ProtocolMembers,
-                              members);
+    nominalType->lookupQualified(nominalType, DeclNameRef(VD->getName()),
+                                 VD->getLoc(), NLOptions::NL_ProtocolMembers,
+                                 members);
```

Comment:
I guess this was also an issue with the previous code, but what if you have a decl that matches by name, but not by type? e.g something like:

```swift
protocol P {}
extension P {
  /// hello
  func foo(_ x: Int) {}
}

struct S: P {
  func foo(_ x: String) {}
}

func bar(_ x: S) {
  x.foo("")
}
```

IMO it could be unexpected to inherit the doc comment in that case.

Reply:
Oh, good idea. Added a test case based on the return type.